[PLT-6656] Migrating sidebar.jsx to redux/PureComponent

jespino opened the pull request:
#### Summary
After splitting the LHS in subcomponents, It migrates the Sidebar component to redux PureComponent.

It depens on mattermost/mattermost-redux#348 and #396 PRs.

#### Ticket Link
[PLT-6656](https://mattermost.atlassian.net/browse/PLT-6656) (It doesn't fix the entire issue)

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features)
- [x] Touches critical sections of the codebase (It touches the channels sidebar)

hmhealey commented on components/sidebar/sidebar_channel/index.js:
```diff
@@ -0,0 +1,81 @@
+        unreadMsgs = 1;
+    }
+
+    let teammate = {}
```
Would you also mind manually running ESLint on this file? There's a bug where it doesn't run on .js files, and this line is missing a semicolon, so it'll definitely trip it up

jespino replied:
Done

hmhealey commented on components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link.jsx:
```diff
@@ -0,0 +1,111 @@
+            element = (
+                <button
+                    className={'btn btn-link ' + this.props.rowClass}
+                    onClick={() => this.handleClick(this.props.link)}
```
Instead of passing `this.props.link` into `this.handleClick`, we can just have the method use `this.props.link` directly so that we aren't creating a new function here

jespino replied:
Done

hmhealey commented on components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link_icon.jsx:
```diff
@@ -0,0 +1,59 @@
+
+    render() {
+        var icon = null;
+        const globeIcon = Constants.GLOBE_ICON_SVG;
```
This is going to conflict with https://github.com/mattermost/mattermost-webapp/pull/453 when that goes in

jespino replied:
I'll fix it when the #453 is merged, or we can solve it in the #453, whatever you want.

hmhealey replied:
That PR has been merged now, so can you use the new GlobeIcon, LockIcon, and ArchiveIcon components instead of importing these from constants.jsx?

jespino replied:
Done!

hmhealey commented on components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link_icon.jsx:
```diff
@@ -0,0 +1,59 @@
+
+import {Constants} from 'utils/constants.jsx';
+
+import StatusIcon from '../../status_icon.jsx';
```
Can you import this as `components/status_icon.jsx` instead of using the relative path?

jespino replied:
Done

hmhealey commented on components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link_close_button.jsx:
```diff
@@ -0,0 +1,70 @@
+    render() {
+        let closeButton = null;
+
+        let removeTooltip = (
```
Could you move the code to set the tooltip into the `if (this.props.handleClose && !this.props.badge)` block to make it clear that that's only used when we have a close handler? I thought there was some new functionality here that I had missed

jespino replied:
Done

hmhealey commented on components/sidebar/sidebar_channel_button_or_link/sidebar_channel_button_or_link_icon.jsx:
```diff
@@ -0,0 +1,59 @@
+                    <StatusIcon
+                        type='avatar'
+                        status={this.props.channelStatus}
+                    />);
```
The `);` should be on its own line

jespino replied:
Done

hmhealey commented:
Looks pretty good. I've got a bunch of little things I noticed, but that's natural for a big PR

jwilander commented on components/sidebar/index.js:
```diff
@@ -0,0 +1,52 @@
+function mapDispatchToProps(dispatch) {
+    return {
+    };
+}
```
If there's no actions, we can just remove this function

jespino replied:
Now I use an action, so now i need it :)

jwilander commented on components/sidebar/sidebar.jsx:
```diff
@@ -0,0 +1,692 @@
+        memberships: PropTypes.object.isRequired,
+        unreads: PropTypes.object.isRequired,
+        isSystemAdmin: PropTypes.bool.isRequired,
+        isTeamAdmin: PropTypes.bool.isRequired
```
Can we add some comments to the props? It's a big component and having some comments on these will make it easier for someone to parse :)

jespino replied:
Done

jwilander commented on components/sidebar/sidebar.jsx:
```diff
@@ -0,0 +1,692 @@
+    }
+
+    render() {
+        this.updateTitle();
```
If we can pull this out and only do it when the props change that would be great. It's very un-React-like to have side effects like this in the render function and it might give us a perf boost if we're rendering more often than just on channel switch

It might be better off if componentDidUpdate() than componentWillReceiveProps(), not sure though

I know this PR is huge already so if we were doing this before then I'm OK with just splitting this (and maybe some other stuff) into improvement tickets

jespino replied:
I have no problem doing this. But I'm not sure if it a side-effect or "not-rendering" thing. It only modifies the document.title which is, in some way, part of the rendering process. We aren't changing the global state, the local state, or any object attribute, it is only reading the props and "drawing" the title based on it.

Move it to the componentDidUpdate is trivial, but I'm not sure is the correct solution.

jespino replied:
I finally moved it to the componentDidUpdate, but I'm not sure is the correct place, as I said, IMHO it is part of the rendering process, other thing is that don't want to "re-render" this part each time.

Other thing is that the location of the title management inside the sidebar is weird, because is something global, to the application, not extrictly dependand on the sidebar. I think can be a Redux component called from the root component. There are other places where is used, but aren't too much and I think we can infer the title from the route + the global state.

jwilander commented on components/sidebar/sidebar_channel/sidebar_channel.jsx:
```diff
@@ -0,0 +1,200 @@
+            savePreferences: PropTypes.func.isRequired,
+            leaveChannel: PropTypes.func.isRequired
+        }).isRequired
+    }
```
Comments here as well would be great

jespino replied:
Done! :)

jwilander commented on routes/route_team.jsx:
```diff
@@ -313,7 +313,7 @@ export default {
                         Promise.all([
                             System.import('components/team_sidebar'),
-                            System.import('components/sidebar.jsx'),
+                            System.import('components/sidebar'),
```
Going to conflict with #501

jespino replied:
This is rebased, so the conflict is already fixed.

jespino commented on components/sidebar/sidebar.jsx:
```diff
@@ -0,0 +1,692 @@
+        const directMessageItems = this.props.directAndGroupChannelIds.map(this.createSidebarChannel);
+
+        // update the favicon to show if there are any notifications
+        if (this.lastBadgesActive !== this.badgesActive) {
```
Yes, I only copied the pattern, but I will try to change it.

jespino replied:
Moved to a function and called on componentDidUpdate.

jespino commented on components/sidebar/sidebar.jsx:
```diff
@@ -0,0 +1,692 @@
+                {moreChannelsModal}
+
+                <SidebarHeader
+                    teamDisplayName={this.props.currentTeam.display_name}
```
I like the idea, but It can be done in other PR, probably in a SidebarHeader migration to redux/pure-component. I think this PR is quite big already, so I prefer to not keep adding complexity.

jespino commented on components/sidebar/sidebar.jsx:
```diff
@@ -0,0 +1,692 @@
+        }
+
+        if (this.props.memberships[channel.id] && this.props.unreads.mentions) {
+            this.badgesActive = true;
```
Moved to a function and called from componentDidUpdate

jespino commented on components/sidebar/sidebar.jsx:
```diff
@@ -0,0 +1,692 @@
+        return this.getDisplayedChannelsForProps(this.props);
+    }
+
+    getDisplayedChannelsForProps = (props) => {
```
Done :)

jespino commented on components/sidebar/sidebar.jsx:
```diff
@@ -0,0 +1,692 @@
+                nextIndex = curIndex - 1;
+            }
+            const nextChannel = allChannelIds[Utils.mod(nextIndex, allChannelIds.length)];
+            browserHistory.push('/' + this.props.currentTeam.name + '/channels/' + this.props.allChannels[nextChannel].name);
```
Done, created a new action "GoToChannelById" and called from the component.

jespino commented on components/sidebar/index.js:
```diff
@@ -0,0 +1,52 @@
+
+function mapStateToProps(state, ownProps) {
+    const config = getConfig(state);
+    const currentChannel = getCurrentChannel(state) || {};
```
Done

jespino commented on components/sidebar/index.js:
```diff
@@ -0,0 +1,52 @@
+    const currentChannel = getCurrentChannel(state) || {};
+
+    return {
+        ...ownProps,
```
Done

jespino commented on components/sidebar/sidebar_channel/index.js:
```diff
@@ -0,0 +1,81 @@
+        unreadMsgs = 1;
+    }
+
+    let teammate = {}
```
Done

jespino commented on components/sidebar/sidebar_channel/index.js:
```diff
@@ -0,0 +1,81 @@
+
+    let teammate = {}
+    if (channel.teammate_id) {
+        teammate = getUser(state, channel.teammate_id) || {};
```
Done

jespino commented on components/sidebar/sidebar_channel/index.js:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    return {
+        ...ownProps,
```
Done

jespino commented:
Now is ready, almost all changes done (other explained), tests passing and styles checked :)

jespino commented:
I also removed the usage of loading in the sidebar channel, because, I almost sure the code is not reachable. There is a sidebar component state named loadingDMChannel which is initialized to -1, and then, never changed and is only compared to the channel index (take from a map method parameter, so it will never be -1). So I removed the usage of the loadingDMChannel in the code and the state key.

jespino commented on components/sidebar/sidebar.jsx:
```diff
@@ -0,0 +1,761 @@
+    setBadgesActiveAndFavicon() {
+        this.lastBadgesActive = this.badgesActive;
+
+        this.getDisplayedChannels().map((channelId) => {
```
Finally I have realized I don't need the map at all, because the check of existence of at least a membership comes from the old code, and I think is not needed to show badge in the title.

jespino commented:
Rebased and fixed the last suggestions from @hmhealey

hmhealey commented:
@jespino This needs another rebase, but you might want to wait on that since it won't be merged until after the release-4.6 branch is cut

jespino commented:
Rebased anyway, I prefer to keep the PR up-to-date as much as possible :)
